researchspace/researchspace
# Add support for websocket protocol to the proxy servlet.

# Why

To be able to proxy through ResearchSpace applications like new 3M UI or Jupiter Notebooks. 

# What

ResearchSpace already have proxy feature implemented with https://github.com/mitre/HTTP-Proxy-Servlet library. This change extends that integration to support websocket protocol using jetty built-in websocket API. 

There are no changes to the existing configuration.

# Video / Gif / Screenshot

Some example of 3M UI and Jupiter Notebooks integration into Thinking Frames:
![Screenshot 2024-04-29 at 21-56-19 X3MLDashboard Artresearch](https://github.com/researchspace/researchspace/assets/2255044/de30dea9-a1f7-416e-a424-7a6a23241daa)


# How To Test

You can use the following `docker-compose.yml` file to setup Jupiter Notebooks locally:
```
version: '3'

services:
  jupiter:
    image: quay.io/jupyter/scipy-notebook:latest
    ports:
      - 10000:8888
    user: root
    command: |
      start-notebook.py --ServerApp.root_dir=/home/jovyan/jupiter --ServerApp.base_url="/proxy/jupiter/" --ServerApp.token="" --ServerApp.password="" --ServerApp.disable_check_xsrf=True --ServerApp.allow_origin="*" --ServerApp.tornado_settings='{"headers":{"Access-Control-Allow-Origin":"*", "Content-Security-Policy":"frame-ancestors * self "}}'
    volumes:
      - ./data:/home/jovyan/jupiter
    environment:
      JUPYTER_ENABLE_LAB: 1
      CHOWN_HOME: 'yes'
      CHOWN_HOME_OPTS: '-R'
```

To test it with dev RS build, add the following config to the `runtime-data/config/proxy.prop` file:

```
config.proxy.jupiter.targetUri=http://localhost:10000/proxy/jupiter
```

When you start the RS instance go to `localhost:10214/proxy/jupiter/lab` in the browesr, you should see Jupiter Notebooks interface, try to create a notebook and execute some simple python function. If you can see the result then the proxy works as expected:
![Screenshot 2024-04-29 at 22-02-17 Untitled ipynb - JupyterLab](https://github.com/researchspace/researchspace/assets/2255044/365aa421-9a8b-4157-bd42-37d006c451f4)

One can use iframe to add it Thinking Frames template:

```
   <template id="jupiterTemplate">
     <div style='flex: 1 1 0px; overflow: auto;'>
       <iframe src="/proxy/jupiter/lab/" scrolling="no" frameborder="0" style="position: relative; height: 100%; width: 100%;"></iframe>
     </div>
   </template>
```
